File: client/bets/protocol/bet_protocol.go
package protocol

import (
	"encoding/binary"
	"net"

	"github.com/7574-sistemas-distribuidos/docker-compose-init/client/bets/models"
)

type betProtocol struct {
	conn net.Conn
}

func NewBetProtocol(serverAddress string) (BetProtocol, error) {
	conn, err := net.Dial("tcp", serverAddress)
	if err != nil {
		return nil, err
	}
	return &betProtocol{
		conn: conn,
	}, nil
}

func (bp *betProtocol) SendAgencyID(agencyID int) error {
	// [HEADER][AGENCY_ID]

	payload := []byte{AGENCY_ID_HEADER} // 1 byte

	agencyIDBytes := make([]byte, INT_SIZE) // 4 bytes
	binary.BigEndian.PutUint32(agencyIDBytes, uint32(agencyID))
	payload = append(payload, agencyIDBytes...)

	if err := bp.writeFull(payload); err != nil {
		return err
	}

	return nil
}

func (bp *betProtocol) SendBetBatch(batch []models.Bet) error {
	// [HEADER][BATCH_SIZE][BET1_LEN][BET1]...[BETN_LEN][BETN]

	payload := []byte{BET_DATA_HEADER} // 1 byte

	batchSize := make([]byte, INT_SIZE) // 4 bytes
	binary.BigEndian.PutUint32(batchSize, uint32(len(batch)))
	payload = append(payload, batchSize...)

	for _, bet := range batch {
		betBytes := []byte(bet.ToString())
		betLen := make([]byte, INT_SIZE) // 4 bytes
		binary.BigEndian.PutUint32(betLen, uint32(len(betBytes)))
		payload = append(payload, betLen...)
		payload = append(payload, betBytes...)
	}

	if err := bp.writeFull(payload); err != nil {
		return err
	}

	return nil
}

func (bp *betProtocol) NotifyBetsFinished() error {
	// [HEADER]
	payload := []byte{BETS_FINISHED_HEADER} // 1 byte

	if err := bp.writeFull(payload); err != nil {
		return err
	}

	return nil
}

func (bp *betProtocol) ReceiveConfirmation() (bool, error) {
	confirmation := make([]byte, PROTOCOL_HEADER_SIZE)

	if err := bp.readFull(confirmation); err != nil {
		return false, err
	}

	return confirmation[0] == BET_CONFIRMATION_HEADER, nil
}

func (bp *betProtocol) ReceiveWinnerIDs() ([]int, error) {
	// [HEADER][IDS_LEN][ID1]...[IDN]

	header := make([]byte, PROTOCOL_HEADER_SIZE)
	if err := bp.readFull(header); err != nil {
		return nil, err
	}
	if header[0] != WINNER_IDS_HEADER {
		return nil, nil // Invalid header
	}

	numWinnersBytes := make([]byte, INT_SIZE)
	if err := bp.readFull(numWinnersBytes); err != nil {
		return nil, err
	}
	numWinners := binary.BigEndian.Uint32(numWinnersBytes)

	winnerIDs := make([]int, numWinners)
	for i := uint32(0); i < numWinners; i++ {
		winnerIDBytes := make([]byte, INT_SIZE)
		if err := bp.readFull(winnerIDBytes); err != nil {
			return nil, err
		}
		winnerIDs[i] = int(binary.BigEndian.Uint32(winnerIDBytes))
	}

	return winnerIDs, nil
}

func (bp *betProtocol) Shutdown() error {
	return bp.conn.Close()
}

/* UTILS */

func (bp *betProtocol) writeFull(data []byte) error {
	totalWritten := 0
	for totalWritten < len(data) {
		n, err := (bp.conn).Write(data[totalWritten:])
		if err != nil {
			return err
		}
		totalWritten += n
	}
	return nil
}

func (bp *betProtocol) readFull(buf []byte) error {
	totalRead := 0
	for totalRead < len(buf) {
		n, err := bp.conn.Read(buf[totalRead:])
		if err != nil {
			return err
		}
		totalRead += n
	}
	return nil
}

File: server/bets/protocol/bet_protocol.py
import socket
import bets.utils as utils
from bets.models.bet import Bet
from bets.protocol.consts import *
from bets.protocol.errors import *

class BetProtocol:
    def __init__(self, skt: socket.socket) -> None:
        self._skt = skt
        self._agency_id = None

    def shutdown(self) -> None:
        """ Closes the connection with the client. """
        if self._skt.fileno() != -1:
            self._skt.shutdown(socket.SHUT_RDWR)
            self._skt.close()

    def receive_agency_id(self) -> int:
        """ Receives the agency ID from client and returns it. """

        # [HEADER][AGENCY_ID]
        header = int.from_bytes(self.__read_full(PROTOCOL_HEADER_SIZE), byteorder='big')
        if header != AGENCY_ID_HEADER:
            raise ValueError("Received invalid header.")

        agency_id_bytes = self.__read_full(INT_SIZE)
        agency_id = int.from_bytes(agency_id_bytes, byteorder='big')

        self._agency_id = agency_id
        return agency_id

    def receive_bet_batch(self) -> list[Bet]:
        """ Receives a batch of bets from client and returns them. """

        if self._agency_id is None:
            raise AgencyIDNotSetError("Agency ID must be set before receiving bets.")

        # [HEADER][BATCH_SIZE][BET1_LEN][BET1]...[BETN_LEN][BETN]
        header = int.from_bytes(self.__read_full(PROTOCOL_HEADER_SIZE), byteorder='big')
        if header == BETS_FINISHED_HEADER:
            return []
        elif header != BET_DATA_HEADER:
            raise ValueError("Received invalid header.")

        batch_size_bytes = self.__read_full(INT_SIZE)
        batch_size = int.from_bytes(batch_size_bytes, byteorder='big')

        bets = []
        for _ in range(batch_size):
            msg_len_bytes = self.__read_full(INT_SIZE)
            msg_len = int.from_bytes(msg_len_bytes, byteorder='big')

            msg_bytes = self.__read_full(msg_len)
            msg_str = msg_bytes.decode('utf-8')

            try:
                first_name, last_name, document, birthdate, number = msg_str.split('|')
                bets.append(Bet(str(self._agency_id), first_name, last_name, document, birthdate, number))
            except ValueError:
                continue # Ignore malformed bets

        if len(bets) != batch_size:
            raise InconsistentBatchSizeError(batch_size)

        return bets

    def send_bet_confirmation(self) -> None:
        """ Sends a bet confirmation to the client. """
        confirmation_code_bytes = BET_CONFIRMATION_HEADER.to_bytes(length=1, byteorder='big')
        self.__write_full(confirmation_code_bytes)

    def send_winner_ids(self, ids: list[int]) -> None:
        """ Sends the list of winner IDs to the client. """

        # [HEADER][IDS_LEN][ID1]...[IDN]
        bytes_to_send = bytearray()

        bytes_to_send.extend(WINNER_IDS_HEADER.to_bytes(length=1, byteorder='big'))
        bytes_to_send.extend(len(ids).to_bytes(length=INT_SIZE, byteorder='big'))
        for id in ids:
            id_bytes = id.to_bytes(length=INT_SIZE, byteorder='big')
            bytes_to_send.extend(id_bytes)

        self.__write_full(bytes(bytes_to_send))
    
    def __read_full(self, n: int) -> bytes:
        buff = bytearray(n)
        view = memoryview(buff)
        bytes_received = 0
        while bytes_received < n:
            chunk_size = self._skt.recv_into(view[bytes_received:])
            if chunk_size == 0:
                raise ConnectionClosedError("Socket connection closed.")
            bytes_received += chunk_size
        return bytes(buff)


    def __write_full(self, data: bytes) -> None:
        view = memoryview(data)
        bytes_sent = 0
        n = len(data)
        while bytes_sent < n:
            chunk_size = 0
            try:
                chunk_size = self._skt.send(view[bytes_sent:])
            except (BrokenPipeError, ConnectionResetError):
                raise ConnectionClosedError("Socket connection closed.")
            bytes_sent += chunk_size
